### ISSUE TITLE
Index Out Of Bounds Exception without relatable error message

### ISSUE THREAD

**Issue by HartmannS** [detekt/detekt#2693]
## Expected Behavior
`./gradlew check` should not run into exception, it should produce meaningful error message

## Observed Behavior
```
Analyzing '/usr/app/src/main/kotlin/simpleTensorFlow/domain/layer/CrossEntropyLayer.kt' led to an exception.
The original exception message was: java.lang.IndexOutOfBoundsException: Wrong offset: 1750. Should be in range: [0, 1749]
Running detekt '1.8.0' on Java '13-ea+32' on OS 'Linux'.
If the exception message does not help, please feel free to create an issue on our GitHub page.
java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:314)
java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:319)
java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1773)
java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
java.base/java.lang.Thread.run(Thread.java:830)
org.jetbrains.kotlin.com.intellij.openapi.editor.impl.LineSet.findLineIndex(LineSet.java:178)
org.jetbrains.kotlin.com.intellij.openapi.editor.impl.DocumentImpl.getLineNumber(DocumentImpl.java:991)
org.jetbrains.kotlin.diagnostics.PsiDiagnosticUtils.offsetToLineAndColumn(PsiDiagnosticUtils.java:73)
org.jetbrains.kotlin.diagnostics.DiagnosticUtils.getLineAndColumnInPsiFile(DiagnosticUtils.java:84)
io.gitlab.arturbosch.detekt.rules.LinesOfCodeKt.line(LinesOfCode.kt:47)
io.gitlab.arturbosch.detekt.rules.LinesOfCodeKt$linesOfCode$1.invoke(LinesOfCode.kt:43)
io.gitlab.arturbosch.detekt.rules.LinesOfCodeKt$linesOfCode$1.invoke(LinesOfCode.kt)
kotlin.sequences.TransformingSequence$iterator$1.next(Sequences.kt:172)
kotlin.sequences.DistinctIterator.computeNext(Sequences.kt:524)
kotlin.collections.AbstractIterator.tryToComputeNext(AbstractIterator.kt:42)
kotlin.collections.AbstractIterator.hasNext(AbstractIterator.kt:29)
kotlin.sequences.SequencesKt___SequencesKt.count(_Sequences.kt:1063)
io.gitlab.arturbosch.detekt.rules.LinesOfCodeKt.linesOfCode(LinesOfCode.kt:45)
io.gitlab.arturbosch.detekt.rules.LinesOfCodeKt.linesOfCode$default(LinesOfCode.kt:42)
io.gitlab.arturbosch.detekt.rules.complexity.LargeClass.visitClassOrObject(LargeClass.kt:59)
org.jetbrains.kotlin.psi.KtVisitorVoid.visitClassOrObject(KtVisitorVoid.java:465)
org.jetbrains.kotlin.psi.KtVisitorVoid.visitClassOrObject(KtVisitorVoid.java:21)
org.jetbrains.kotlin.psi.KtVisitor.visitClass(KtVisitor.java:33)
org.jetbrains.kotlin.psi.KtVisitorVoid.visitClass(KtVisitorVoid.java:33)
org.jetbrains.kotlin.psi.KtVisitorVoid.visitClass(KtVisitorVoid.java:459)
org.jetbrains.kotlin.psi.KtVisitorVoid.visitClass(KtVisitorVoid.java:21)
org.jetbrains.kotlin.psi.KtClass.accept(KtClass.kt:20)
org.jetbrains.kotlin.psi.KtElementImplStub.accept(KtElementImplStub.java:59)
org.jetbrains.kotlin.com.intellij.psi.impl.source.tree.SharedImplUtil.acceptChildren(SharedImplUtil.java:200)
org.jetbrains.kotlin.com.intellij.psi.impl.source.PsiFileImpl.acceptChildren(PsiFileImpl.java:733)
org.jetbrains.kotlin.psi.KtTreeVisitorVoid.visitElement(KtTreeVisitorVoid.java:25)
org.jetbrains.kotlin.com.intellij.psi.PsiElementVisitor.visitFile(PsiElementVisitor.java:34)
org.jetbrains.kotlin.psi.KtVisitor.visitKtFile(KtVisitor.java:73)
org.jetbrains.kotlin.psi.KtVisitorVoid.visitKtFile(KtVisitorVoid.java:69)
org.jetbrains.kotlin.psi.KtVisitorVoid.visitKtFile(KtVisitorVoid.java:513)
org.jetbrains.kotlin.psi.KtVisitorVoid.visitKtFile(KtVisitorVoid.java:21)
org.jetbrains.kotlin.psi.KtFile.accept(KtFile.kt:242)
org.jetbrains.kotlin.psi.KtFile.accept(KtFile.kt:229)
io.gitlab.arturbosch.detekt.api.internal.BaseRule.visit(BaseRule.kt:41)
io.gitlab.arturbosch.detekt.api.internal.BaseRule.visitFile(BaseRule.kt:31)
io.gitlab.arturbosch.detekt.core.Detektor$analyze$4.invoke(Detektor.kt:101)
io.gitlab.arturbosch.detekt.core.Detektor.analyze(Detektor.kt:112)
io.gitlab.arturbosch.detekt.core.Detektor.access$analyze(Detektor.kt:19)
io.gitlab.arturbosch.detekt.core.Detektor$runAsync$$inlined$map$lambda$1.invoke(Detektor.kt:68)
io.gitlab.arturbosch.detekt.core.Detektor$runAsync$$inlined$map$lambda$1.invoke(Detektor.kt:19)
io.gitlab.arturbosch.detekt.core.TaskPoolKt$task$1.get(TaskPool.kt:13)
java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1771)
java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
java.base/java.lang.Thread.run(Thread.java:830)

```
```

Analyzing '/usr/app/src/main/kotlin/simpleTensorFlow/domain/layer/FullyConnectedLayer.kt' led to an exception.
The original exception message was: java.lang.IndexOutOfBoundsException: Wrong offset: 2786. Should be in range: [0, 2785]
Running detekt '1.8.0' on Java '13-ea+32' on OS 'Linux'.
If the exception message does not help, please feel free to create an issue on our GitHub page.
java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:314)
java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:319)
java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1773)
java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
java.base/java.lang.Thread.run(Thread.java:830)
org.jetbrains.kotlin.com.intellij.openapi.editor.impl.LineSet.findLineIndex(LineSet.java:178)
org.jetbrains.kotlin.com.intellij.openapi.editor.impl.DocumentImpl.getLineNumber(DocumentImpl.java:991)
org.jetbrains.kotlin.diagnostics.PsiDiagnosticUtils.offsetToLineAndColumn(PsiDiagnosticUtils.java:73)
org.jetbrains.kotlin.diagnostics.DiagnosticUtils.getLineAndColumnInPsiFile(DiagnosticUtils.java:84)
io.gitlab.arturbosch.detekt.rules.LinesOfCodeKt.line(LinesOfCode.kt:47)
io.gitlab.arturbosch.detekt.rules.LinesOfCodeKt$linesOfCode$1.invoke(LinesOfCode.kt:43)
io.gitlab.arturbosch.detekt.rules.LinesOfCodeKt$linesOfCode$1.invoke(LinesOfCode.kt)
kotlin.sequences.TransformingSequence$iterator$1.next(Sequences.kt:172)
kotlin.sequences.DistinctIterator.computeNext(Sequences.kt:524)
kotlin.collections.AbstractIterator.tryToComputeNext(AbstractIterator.kt:42)
kotlin.collections.AbstractIterator.hasNext(AbstractIterator.kt:29)
kotlin.sequences.SequencesKt___SequencesKt.count(_Sequences.kt:1063)
io.gitlab.arturbosch.detekt.rules.LinesOfCodeKt.linesOfCode(LinesOfCode.kt:45)
io.gitlab.arturbosch.detekt.rules.LinesOfCodeKt.linesOfCode$default(LinesOfCode.kt:42)
io.gitlab.arturbosch.detekt.rules.complexity.LargeClass.visitClassOrObject(LargeClass.kt:59)
org.jetbrains.kotlin.psi.KtVisitorVoid.visitClassOrObject(KtVisitorVoid.java:465)
org.jetbrains.kotlin.psi.KtVisitorVoid.visitClassOrObject(KtVisitorVoid.java:21)
org.jetbrains.kotlin.psi.KtVisitor.visitClass(KtVisitor.java:33)
org.jetbrains.kotlin.psi.KtVisitorVoid.visitClass(KtVisitorVoid.java:33)
org.jetbrains.kotlin.psi.KtVisitorVoid.visitClass(KtVisitorVoid.java:459)
org.jetbrains.kotlin.psi.KtVisitorVoid.visitClass(KtVisitorVoid.java:21)
org.jetbrains.kotlin.psi.KtClass.accept(KtClass.kt:20)
org.jetbrains.kotlin.psi.KtElementImplStub.accept(KtElementImplStub.java:59)
org.jetbrains.kotlin.com.intellij.psi.impl.source.tree.SharedImplUtil.acceptChildren(SharedImplUtil.java:200)
org.jetbrains.kotlin.com.intellij.psi.impl.source.PsiFileImpl.acceptChildren(PsiFileImpl.java:733)
org.jetbrains.kotlin.psi.KtTreeVisitorVoid.visitElement(KtTreeVisitorVoid.java:25)
org.jetbrains.kotlin.com.intellij.psi.PsiElementVisitor.visitFile(PsiElementVisitor.java:34)
org.jetbrains.kotlin.psi.KtVisitor.visitKtFile(KtVisitor.java:73)
org.jetbrains.kotlin.psi.KtVisitorVoid.visitKtFile(KtVisitorVoid.java:69)
org.jetbrains.kotlin.psi.KtVisitorVoid.visitKtFile(KtVisitorVoid.java:513)
org.jetbrains.kotlin.psi.KtVisitorVoid.visitKtFile(KtVisitorVoid.java:2

```

## Steps to Reproduce
If needed, i can poste src code, gradle settings, properties and all
**Error is gone after `./gradlew ktlintFormat`**

```
tasks{
    withType<Detekt> {
        // Target version of the generated JVM bytecode. It is used for type resolution.
        jvmTarget = "13"
    }
}

detekt {
    parallel = true
    autoCorrect = true
    failFast = true // fail build on any finding
    buildUponDefaultConfig = true // preconfigure defaults
    config = files("$projectDir/config/detekt/detekt.yml") // point to your custom config defining rules to run, overwriting default behavior
    baseline = file("$projectDir/config/detekt/baseline.xml") // a way of suppressing issues before introducing detekt

    reports {
        html.enabled = true // observe findings in your browser with structure and code snippets
        xml.enabled = true // checkstyle like format mainly for integrations like Jenkins
        txt.enabled = true // similar to the console output, contains issue signature to manually edit baseline files
    }
}
```

## Your Environment
Is repeatable with docker image openjdk:13-alpine and graldew 5.4.0
detekt-plugin detekt-formatting is used

**Comment by BraisGabin** [comment]
Hi @HartmannS, I was not able to reproduce it. Could you provide a project/configuration/file (I'm not sure which is the important piece here to reproduce the error) so I can dig in?

**Comment by HartmannS** [comment]
Hey, @BraisGabin,
i could not reproduce the issue with my copy of the repo.
Even in a fixed docker container, the exception vanished.
If you would like to check it out for yourself, i can upload a copy of the repo, perhaps there was a minor version bump of something, like jdk, jvm or base image version, which caused the issue.
Else, the issue can be closed, thank you for your time.

**Comment by BraisGabin** [comment]
OK, I close it. If the issue appears later, please, reopen it :).

**Comment by andrey-bolduzev** [comment]
Just hit the same issue with v0.15.0.

```
* What went wrong:
Execution failed for task ':detektAll'.
> java.lang.IllegalStateException: Analyzing <path>/app/src/test/java/MockUtils.kt led to an exception. 
  The original exception message was: java.lang.IndexOutOfBoundsException: Wrong offset: 834. Should be in range: [0, 833]
  Running detekt '1.15.0' on Java '14.0.2+12' on OS 'Linux'
  If the exception message does not help, please feel free to create an issue on our GitHub page.

* Try:
Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Exception is:
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':detektAll'.
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.lambda$executeIfValid$3(ExecuteActionsTaskExecuter.java:186)
        at org.gradle.internal.Try$Failure.ifSuccessfulOrElse(Try.java:268)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeIfValid(ExecuteActionsTaskExecuter.java:184)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:173)
        at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:109)
        at org.gradle.api.internal.tasks.execution.FinalizePropertiesTaskExecuter.execute(FinalizePropertiesTaskExecuter.java:46)
        at org.gradle.api.internal.tasks.execution.ResolveTaskExecutionModeExecuter.execute(ResolveTaskExecutionModeExecuter.java:51)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:57)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:56)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:36)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.executeTask(EventFiringTaskExecuter.java:77)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:55)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:52)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:200)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:195)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:75)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:68)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:153)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:68)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.call(DefaultBuildOperationRunner.java:62)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.lambda$call$2(DefaultBuildOperationExecutor.java:76)
        at org.gradle.internal.operations.UnmanagedBuildOperationWrapper.callWithUnmanagedSupport(UnmanagedBuildOperationWrapper.java:54)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:76)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:52)
        at org.gradle.execution.plan.LocalTaskNodeExecutor.execute(LocalTaskNodeExecutor.java:41)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:411)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:398)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:391)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:377)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.lambda$run$0(DefaultPlanExecutor.java:127)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.execute(DefaultPlanExecutor.java:191)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.executeNextNode(DefaultPlanExecutor.java:182)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.run(DefaultPlanExecutor.java:124)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
```

The file contents are as follows (the part that causes the failure):

```kotlin
import com.nhaarman.mockitokotlin2.stub
import org.mockito.internal.invocation.InterceptedInvocation
import org.mockito.invocation.InvocationOnMock
import kotlin.coroutines.Continuation
import kotlin.coroutines.intrinsics.startCoroutineUninterceptedOrReturn

@Suppress("UNCHECKED_CAST")
fun <T> (suspend () -> T).stubAnswer(answer: suspend (InvocationOnMock) -> T?) {
    stub {
        onBlocking { this@stubAnswer() }.thenAnswer {
            //all suspend functions/lambdas has Continuation as the last argument.
            //InvocationOnMock does not see last argument
            val rawInvocation = it as InterceptedInvocation
            val continuation = rawInvocation.rawArguments.last() as Continuation<T?>

            answer.startCoroutineUninterceptedOrReturn(it, continuation)
        }
    }
}
```

with:

```kotlin
implementation("org.jetbrains.kotlinx:kotlinx-coroutines-core:1.4.2")
testImplementation("com.nhaarman.mockitokotlin2:mockito-kotlin:2.2.0")
```

Duplicated the file between two different projects, the issue persists.

**Comment by chao2zhang** [comment]
@andrey-bolduzev I tried reproducing it with Java 11.
I was wondering if this is impacted by the java/jvm version `14.0.2+12`. Is it possible to get a build scan from your failed runs?

**Comment by andrey-bolduzev** [comment]
> @andrey-bolduzev I tried reproducing it with Java 11.
> I was wondering if this is impacted by the java/jvm version `14.0.2+12`. Is it possible to get a build scan from your failed runs?

Thank you for your quick reply!

I've tried it with Java 11 as well for the exact same error.

I've created a stripped-down version of a project where the issue can be reproduced by running `./gradlew detektAll`: https://github.com/andrey-bolduzev/detekt_issue_2693.

Here's a build scan for it: https://scans.gradle.com/s/6upaj4nydwouo.

**Comment by chao2zhang** [comment]
I was able to reproduce and observed the following behavior:

- KtFile.text.length = 834
- KtFile.viewProvider.document.textLength = 832

It looks like a bug in the Kotlin compiler and I will try reporting this bug to Kotlin. At the same time, I am also exploring options to refactor us off using `DiagnosticUtils.getLineAndColumnInPsiFile()` the calculate line counts of a KtElement excluding comments.

**Comment by andrey-bolduzev** [comment]
Ah, your comment made me realize the issue is related to the 2 comment lines with missing leading spaces. When the inspection is not triggered, the crash doesn't happen. At least for now I can just fix those 2 lines and not exclude the whole file like I did before. Thank you!

**Comment by chao2zhang** [comment]
This is a super interesting edge case where we execute ktlint rules first with `autoCorrect = true`, so the content of `KtFile` is modified (length from 834 to 832), then execute detekt rules like `LongMethod`, which raises the issue.

**Comment by BraisGabin** [comment]
🎉🎉 @andrey-bolduzev thanks for the reproducible project and the final hint to find the root issue. And great job @chao2zhang to find that root problem!

**Comment by chao2zhang** [comment]
I am working on a short-term patch fix for this issue. I am running into the exact same exception when upgrading to detekt 1.16-RC and there are "tons" of `Indentation` violation in ktlint that I strongly believe my colleagues will yell at me if  autocorrect crashes.

**Comment by schalkms** [comment]
@chao2zhang does PR #3446 completely fix this problem, so that we can close this issue?

**Comment by chao2zhang** [comment]
Yes it should have. Closing this issue.
